packages/roosterjs-editor-core/lib/undo/Undo.ts

```diff
@@ -0,0 +1,226 @@
+import UndoSnapshots from './UndoSnapshots';
+import containsImage from './containsImage';
+import {
+    ContentChangedEvent,
+    PluginDomEvent,
+    PluginEvent,
+    PluginEventType,
+} from 'roosterjs-editor-types';
+import Editor from '../editor/Editor';
+import UndoService from '../editor/UndoService';
+import { buildSnapshot, restoreSnapshot} from './snapshotUtils';
+
+const KEY_BACKSPACE = 8;
+const KEY_DELETE = 46;
+const KEY_SPACE = 32;
+const KEY_ENTER = 13;
+const KEY_PAGEUP = 33;
+const KEY_DOWN = 40;
+
+/**
+ * Provides snapshot based undo service for Editor
+ */
+export default class Undo implements UndoService {
+    private editor: Editor;
+    private isRestoring: boolean;
+    private hasNewContent: boolean;
+    private undoSnapshots: UndoSnapshots;
+    private lastKeyPress: number;
+
+    /**
+     * Create an instance of Undo
+     * @param preserveSnapshots True to preserve the snapshots after dispose, this allows
+     * this object to be reused when editor is disposed and created again
+     */
+    constructor(
+        private preserveSnapshots?: boolean) {}
+
+    /**
+     * Initialize this plugin. This should only be called from Editor
+     * @param editor Editor instance
+     */
+    public initialize(editor: Editor): void {
+        this.editor = editor;
+
+        // Add an initial snapshot if snapshotsManager isn't created yet
+        if (!this.undoSnapshots) {
+            this.addUndoSnapshot();
+        }
+    }
+
+    /**
+     * Dispose this plugin
+     */
+    public dispose() {
+        this.editor = null;
+
+        if (!this.preserveSnapshots) {
```

Comment:
Don't really understand how this works.. The next time an editor instance is created, will it still use the same Undo instance? If user compose a new email, would he be seeing the old undo snapshots?

Reply:
This is on purpose. It is for reusing undo snapshots. For the same session with different editor instances, we share the undo snapshots. For different session, we should create new Undo object.